README
Sjitter provides, in a simple command line interface, 
a tool to measure jitter, one way delay and bandwidth.

Sjitter is a simple network performance tool to measure:
- jitter (based on the RFC 1889).
- one way delay (based on UDP datagram transit time).
- bandwidth (steel experimental).
Sjitter is a client/server program. Both client and server 
need to be date and time synchronize.


NEWS
-------------
Release notes 
-------------

Version 0.15:
- Package generate with automake

Version 0.14:
- Use timer (-t option)
- Rewrite progress bar procedure
- Man pages

Version 0.13:
- add buffer size parameters (-w option)
- Fix a bug in the minimum value
- Send a SJITTER-STOP datagram when a client is CTRL-C (thanks to Daitau Aaron)

Version 0.12:
- Use DNS name additionnaly to IP address (OK 06/07/2006)
- Multiple measure on the server
- Bug fix

Version 0.11:
- Correct bug in the archive

Version 0.1:
- First beta version


jitter.c
/*
 * Simple Jitter program
 * Alcatel Alenia Space - 2006
 * GPL Licence
 * Author: Nicolas Hennion
 */
 
#include "jitter.h"

/* Global variables */
extern int socketfd;
extern struct sockaddr_in serveraddr;

/* Signals */
void onsignal(int code) {	
	char *buf;
	socklen_t slen=sizeof(serveraddr);

	if ((code == SIGINT) || (code == SIGTERM) || (code == SIGQUIT)) {				
  		/* Send "end" datagram to the server */
  		buf = (char *) malloc(DEFAULT_BUFFER_SIZE);
		snprintf(buf, DEFAULT_BUFFER_SIZE, "SJITTER-END ");
		if (sendto(socketfd, buf, DEFAULT_BUFFER_SIZE, 0, (const struct sockaddr *) &serveraddr, slen) == -1)
			errorexit("sendto");
		/* Close the socket */
		perror("\nProcess killed on user request...");
		close(socketfd);
		exit(1);
	}
	signal(code,onsignal);
}    

/* Display the error and exit */
void errorexit(char *s) {
	perror(s);
	close(socketfd);
	exit(1);
}

/* get the current time ms */
double getcurrenttimems() {
	struct timeval now;

	gettimeofday(&now, NULL);
	
	return (double) (now.tv_sec+now.tv_usec/1.0E6);
}

/* Displat a (text) progressbar on STDOUT */
void progressbar(int pbcurrent, int pbmax) {
	int i, cursor;
	char bar[BUFFER_CHAR];
	
	cursor = (int) (pbcurrent*80/pbmax);
	for (i=0; i<80; i++) {
		if (i>cursor)
			bar[i] = '_';
		else
			bar[i] = '#';
	}
	bar[i] = '\0';	
	printf("\r%s", bar);
	fflush(stdout);
}


jitter.h
/*
 * Simple Jitter program
 * Alcatel Alenia Space - 2006
 * GPL Licence
 * Author: Nicolas Hennion
 */
 
#ifndef JITTER_H_
#define JITTER_H_

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
 
/* Constants */
#define SJITTER_NAME			"SJitter"
#define SJITTER_VERSION			"0.18.1"
#define SJITTERC_NAME			"SJitter client"
#define SJITTERC_VERSION		"0.18.1"
#define SJITTERS_NAME			"SJitter server"
#define SJITTERS_VERSION		"0.18.1"
#define BUFFER_CHAR				1024
#define BUFFER_INET4_ADDR		15

/* Network constants */
#define DEFAULT_BUFFER_SIZE 	1400	/* bytes */
#define MIN_BUFFER_SIZE 		128
#define MAX_BUFFER_SIZE 		8000
#define DEFAULT_PORT_NUMBER 	9930
#define MAX_PORT_NUMBER 		65535
#define DEFAULT_PACKET_NUMBER 	100
#define MAX_PACKET_NUMBER 		65535
#define DEFAULT_TIME_NUMBER 	10
#define DEFAULT_BITRATE			100		/* Kbits per second */
#define DEFAULT_TOS				0x00

/* Headers */
void onsignal(int code);
void errorexit(char *s);
double getcurrenttimems();
void progressbar(int pbcurrent, int pbmax);

#endif /*JITTER_H_*/


sjitterc.c
/*
 * Simple Jitter program (client side)
 * Alcatel Alenia Space - 2006
 * GPL Licence
 * Author: Nicolas Hennion
 */
 
#include "jitter.h"

/* Global variables */
int socketfd; 
struct sockaddr_in serveraddr;
#ifdef IPV6_SUPPORT
struct sockaddr_in6 serveraddr6;
#endif

int main(int argc, char **argv) {
	short int tag6 = 0;
	
	socklen_t slen=sizeof(serveraddr);
#ifdef IPV6_SUPPORT
	socklen_t slen6=sizeof(serveraddr6);	
#endif
	
	struct hostent *hptr;
	char *buf, server_name[BUFFER_CHAR], server_ip[BUFFER_CHAR];
	extern char *optarg;
	int i, errflag=1, n_flag=0, t_flag=0;
	int packet_number=DEFAULT_PACKET_NUMBER, time_number=DEFAULT_TIME_NUMBER;
	int port_number=DEFAULT_PORT_NUMBER, buffer_size=DEFAULT_BUFFER_SIZE, bitrate=DEFAULT_BITRATE, tos = DEFAULT_TOS;
	double begin=0, end=0, sendbegin=0, sendend=0, lastprogressbar=0;
	unsigned long packet_delay, packet_size;
	
	/* Manage arguments */
	#ifdef IPV6_SUPPORT
	while ((i = getopt(argc, argv, "6c:n:t:p:w:b:s:v")) != EOF) {
	#else
	while ((i = getopt(argc, argv, "c:n:t:p:w:b:s:v")) != EOF) {
	#endif	 
		switch (i) { 
			#ifdef IPV6_SUPPORT
			case '6':
				tag6 = 1;
				break;
			#endif	 			
			case 'c': 				
				strncpy(server_name, optarg, BUFFER_CHAR);
				if (!tag6) {
					/* IPv4 */		
					hptr = gethostbyname(server_name);
				#ifdef IPV6_SUPPORT
				} else {
					/* IPv6 */
					hptr = gethostbyname2(server_name,AF_INET6);
				#endif
				}
				if (hptr == NULL) {
					fprintf(stderr, "Invalid server IP address or name...\n"); 					
					errflag = 1;
				} else {					
					inet_ntop(hptr->h_addrtype, *(hptr->h_addr_list), server_ip, sizeof(server_ip));
					errflag = 0;
				}
				break;
			case 's': 		
                /* TOS option specified */
				sscanf(optarg, "%x", &tos);
				if ((tos < 0x00) || (tos > 0xFF))
					errflag = 1;					
				break;
			case 'n':
				sscanf(optarg, "%d", &packet_number);
				n_flag = 1;
				if ((packet_number < 2) || (packet_number > MAX_PACKET_NUMBER) || t_flag)
					errflag = 1;					
				break;				
			case 't':
				sscanf(optarg, "%d", &time_number);
				t_flag = 1;				
				if ((time_number < 2) || n_flag)
					errflag = 1;
				break;				
			case 'p':
				sscanf(optarg, "%d", &port_number);
				if ((port_number < 1024) || (port_number > MAX_PORT_NUMBER))
					errflag = 1;
				break;
			case 'w':
				sscanf(optarg, "%d", &buffer_size);
				if ((buffer_size < MIN_BUFFER_SIZE) || (buffer_size > MAX_BUFFER_SIZE))
					errflag = 1;
				break;
			case 'b':
				sscanf(optarg, "%d", &bitrate);
				if (bitrate < 11 )
					errflag = 1;
				break;
			case 'v':
				fprintf(stderr, "version: sjitter client version %s\n", SJITTERC_VERSION); 
				errflag = 1;
				break;
		} 
	} 
	if (errflag) { 
		#ifdef IPV6_SUPPORT
		fprintf(stderr, "usage: sjitterc [-6] -c SERVER [[-n NBPCKT] | [-t SECOND]] [-p PORTNB] [-w SIZE] [-b BITRAT] [-s TOS]\n");
		fprintf(stderr, "\t-6 : Use the IPv6 protocol\n");
		#else
		fprintf(stderr, "usage: sjitterc -c SERVER [[-n NBPCKT] | [-t SECOND]] [-p PORTNB] [-w SIZE] [-b BITRAT] [-s TOS]\n");		
		#endif 
		fprintf(stderr, "\t-c SERVER: where SERVER is the server IP address or name\n");
		fprintf(stderr, "\t-n NBPCKT: where NCPCKT is the number of datagram (>1 , <%d) [default:%d]\n", MAX_PACKET_NUMBER, DEFAULT_PACKET_NUMBER);
		fprintf(stderr, "\t-t SECOND: where SECOND is the number of second (>1) [default:%d]\n", DEFAULT_TIME_NUMBER);		
		fprintf(stderr, "\t-p PORTNB: where PORTNB is the port number (>1024, <%d) [default:%d]\n", MAX_PORT_NUMBER, DEFAULT_PORT_NUMBER);
		fprintf(stderr, "\t-w SIZE: where SIZE is the application buffer size (bytes) (>%d, <%d) [default:%d]\n", MIN_BUFFER_SIZE, MAX_BUFFER_SIZE, DEFAULT_BUFFER_SIZE);
		fprintf(stderr, "\t-b BITRATE: where BITRATE is the bitrate (IP level) in Kbps  (>10) [default:%d]\n", DEFAULT_BITRATE);	
		fprintf(stderr, "\t-s TOS: where TOS is the hexadecimal value for IP header TOS field (>=0x00, <=0xFF) [default:%x]\n", DEFAULT_TOS);
		if (n_flag && t_flag) {
			fprintf(stderr, "It is not possible to use -n and -t in the same command line\n");
		}
		exit(2);
	}

	/* Buffer allocation */
	buf = (char *) malloc(buffer_size);

	/* Signals management */
	signal(SIGINT, onsignal);
	signal(SIGTERM, onsignal);
	signal(SIGQUIT, onsignal);

	/* Build the socket */
	if (!tag6) {
		/* IPv4 */	
		if ((socketfd=socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP))==-1)
			errorexit("IPv4 socket error");
		/* Set TOS field */
		if (setsockopt(socketfd, IPPROTO_IP, IP_TOS, (char*) &tos, sizeof(tos)) == -1)
           errorexit("IPv4 setsockopt error");
		memset((char *) &serveraddr, sizeof(serveraddr), 0);
		serveraddr.sin_family = AF_INET;
		serveraddr.sin_port = htons(port_number);
		if (inet_aton(server_ip, &serveraddr.sin_addr)==0)
			perror("IPv4 inet_aton error");
	#ifdef IPV6_SUPPORT
	} else {
		/* IPV6 */
		if ((socketfd=socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP))==-1)
			errorexit("IPv6 socket error");
		memset((char *) &serveraddr6, sizeof(serveraddr6), 0);
		serveraddr6.sin6_family = AF_INET6;
		serveraddr6.sin6_port = htons(port_number);
		if (inet_pton(AF_INET6, server_ip, &serveraddr6.sin6_addr)==0)
			perror("IPV6 inet_aton error");
	#endif
	}
	
	if (t_flag) {
		/* -t option set, compute the packet number */
		packet_number = (time_number*(bitrate*1000.0))/((buffer_size+28)*8);
	} else {
		/* Compute the estimate time */
		time_number = (packet_number*((buffer_size+28)*8))/(bitrate*1000.0);
	}
	if (!tag6) {
		/* IPv4 */		
		printf("Send data (%d datagrams of %d bytes / %d Kbps) to the server:port %s:%d\n", 
				packet_number, buffer_size, bitrate, inet_ntoa(serveraddr.sin_addr), port_number);
	#ifdef IPV6_SUPPORT
	} else {
		/* IPv6 */
		char buf6[INET6_ADDRSTRLEN];			
		printf("Send data (%d datagrams of %d bytes / %d Kbps) to the [server]:port [%s]:%d\n", 
				packet_number, buffer_size, bitrate, inet_ntop(AF_INET6, &serveraddr6.sin6_addr, buf6, sizeof(buf6)), port_number);
	#endif
	}
	
	/* Send configuration datagram to the server */
	snprintf(buf, buffer_size, "SJITTER-START %d %d ", packet_number, buffer_size);
	if (!tag6) {
		/* IPv4 */		
		if (sendto(socketfd, buf, buffer_size, 0, (const struct sockaddr *) &serveraddr, slen)==-1)
			errorexit("IPv4 sendto error");
	#ifdef IPV6_SUPPORT
	} else {
		/* IPv6 */
		if (sendto(socketfd, buf, buffer_size, 0, (const struct sockaddr *) &serveraddr6, slen6)==-1)
			errorexit("IPv6 sendto error");		
	#endif
	}

	/* Compute delay for bitrate (goal on the IP level) */
	packet_size = (unsigned long) ((buffer_size+28)*8);
	packet_delay = (unsigned long) ((packet_size/(bitrate*1000.0))*1000000);
	
	/* Send data to the server */
	/* Init the packet number tag */
	i=0;
	/* Star the main loop */
	printf("Sending data (estimate time: %d seconds)...\n", time_number);
	begin = lastprogressbar = getcurrenttimems();
	while (i<packet_number) {
		
		/* Manage bitrate */
		usleep(packet_delay-((sendend-sendbegin)*1000000));
		
		/* Send datagram on the network interface */
		sendbegin = getcurrenttimems();	/* Send begin... */				
		snprintf(buf, buffer_size, "SJITTER-DATA %d %lf ", i, sendbegin);
		if (!tag6) {
			/* IPv4 */				
			if (sendto(socketfd, buf, buffer_size, 0, (const struct sockaddr *) &serveraddr, slen)==-1)
				errorexit("IPv4 sendto error");
		#ifdef IPV6_SUPPORT
		} else {
			/* IPv6 */
			if (sendto(socketfd, buf, buffer_size, 0, (const struct sockaddr *) &serveraddr6, slen6)==-1)
				errorexit("IPv6 sendto error");			
		#endif
		}				
		sendend = getcurrenttimems(); /* ... send end */
		
		/* Display progress bar (every 1 second)*/
		if ((sendend-lastprogressbar) > 1) {
			progressbar(i, packet_number);
			lastprogressbar = sendend;
		}
		
		/* Next packet */
		i++;
	}
	end = getcurrenttimems();

	/* Send "SJITTER-END" datagram to the server */
	snprintf(buf, buffer_size, "SJITTER-END ");
	if (!tag6) {
		/* IPv4 */				
		if (sendto(socketfd, buf, buffer_size, 0, (const struct sockaddr *) &serveraddr, slen)==-1)
			errorexit("IPv4 sendto error");
	#ifdef IPV6_SUPPORT
	} else {
		/* IPv6 */
		if (sendto(socketfd, buf, buffer_size, 0, (const struct sockaddr *) &serveraddr6, slen6)==-1)
			errorexit("IPv6 sendto error");			
	#endif
	}				
		
	/* Summary */
	printf("\r                                                                                \r");
	printf("Summary: %d datagrams sent in %.2lf seconds (%.0lf Kbps)\n", 
		   packet_number, end-begin, (packet_number*buffer_size*8)/(end-begin)/1000);
	
	close(socketfd);
	free(buf);
	
	return 0;
}


sjitters.c
/*
 * Simple Jitter program (server side)
 * Alcatel Alenia Space - 2006
 * GPL Licence
 * Author: Nicolas Hennion
 */
 
#include "jitter.h"

/* Global variables */
int socketfd; 
struct sockaddr_in serveraddr;
struct sockaddr_in6 serveraddr6;

int main(int argc, char **argv) {
	short int tag6 = 0;
	
	struct sockaddr_in clientaddr;
	socklen_t slen=sizeof(clientaddr);
	#ifdef IPV6_SUPPORT 
	socklen_t slen6=sizeof(clientaddr6);
	struct sockaddr_in6 clientaddr6;
	#endif
		
	char *buf, bufstart[MAX_BUFFER_SIZE], bufchar[BUFFER_CHAR];
	extern char *optarg;
	int i, pn, pid;
	int packet_number, buffer_size, port_number=DEFAULT_PORT_NUMBER;
	int errflag=0, iflag=0, checkconf=0, endflag=0;
	double begin=0, end=0, t_send, t_receive, t_transit, t_transitlast, t_transitdelta;
	double jitterlast=0, jittermin=0, jittermax=0, jitteravg=0, jittersum=0, jitterlist[MAX_PACKET_NUMBER];	
	double onewaydelaylast=0, onewaydelaymin=0, onewaydelaymax=0, onewaydelayavg=0, onewaydelaysum=0, onewaydelaylist[MAX_PACKET_NUMBER];	

	/* Manage arguments */
	while ((i = getopt(argc, argv, "6ip:vh")) != EOF) { 
		switch (i) { 
			case '6':
				tag6 = 1;
				break;
			case 'i':
				iflag++;
				break;
			case 'p':
				sscanf(optarg, "%d", &port_number);
				if ((port_number < 1024) || (port_number > MAX_PORT_NUMBER))
					errflag = 1;
				break;
			case 'v':
				fprintf(stderr, "version: sjitter client version %s\n", SJITTERC_VERSION); 
				errflag = 1;
				break;
			case 'h':
				errflag = 1;
				break;
		} 
	} 
	if (errflag) {
		#ifdef IPV6_SUPPORT 
		fprintf(stderr, "usage: sjitters [-6] [-i] [-p PORT]\n"); 
		fprintf(stderr, "\t-6 : Use IPv6 protocol\n");
		#else
		fprintf(stderr, "usage: sjitters [-i] [-p PORT]\n");
		#endif		
		fprintf(stderr, "\t-p PORT: where PORT is the port number (>1024, <%d) [default:%d]\n", MAX_PORT_NUMBER, DEFAULT_PORT_NUMBER);		
		fprintf(stderr, "\t-i : Verbose mode\n");
		exit(2);
	}

	/* Signals management */
	signal(SIGINT,onsignal);
	signal(SIGTERM,onsignal);
	signal(SIGQUIT,onsignal);	

	/* Build the socket */
	if (!tag6) {
		/* IPv4 */
		if ((socketfd=socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP))==-1)
			errorexit("IPv4 socket error");
		memset((char *) &serveraddr, sizeof(serveraddr), 0);
		serveraddr.sin_family = AF_INET;
		serveraddr.sin_port = htons(port_number);
		serveraddr.sin_addr.s_addr = htonl(INADDR_ANY);
		if (bind(socketfd, (struct sockaddr *) &serveraddr, sizeof(serveraddr))==-1)
			errorexit("IPv4 bind error");
	#ifdef IPV6_SUPPORT
	} else {
		/* IPv6 */
		if ((socketfd=socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP))==-1)
			errorexit("IPv6 socket error");	
		memset((char *) &serveraddr6, sizeof(serveraddr6), 0);
		serveraddr6.sin6_family = AF_INET6;
		serveraddr6.sin6_port = htons(port_number);
		serveraddr6.sin6_addr = in6addr_any;
		if (bind(socketfd, (struct sockaddr *) &serveraddr6, sizeof(serveraddr6))==-1)
			errorexit("IPv6 bind error");
	#endif
	}

	/* Main loop */
	while (1) {
		
		/* Wait START (configuration) datagram */
		while (checkconf==0) {
			if (!tag6) {
				/* IPv4 */			
				if (recvfrom(socketfd, bufstart, sizeof(bufstart), 0, (struct sockaddr *) &clientaddr, &slen)==-1)
					errorexit("IPv4 recvfrom error");
			#ifdef IPV6_SUPPORT
			} else {
				/* IPv6 */			
				if (recvfrom(socketfd, bufstart, sizeof(bufstart), 0, (struct sockaddr *) &clientaddr6, &slen6)==-1)
					errorexit("IPv6 recvfrom error");
			#endif
			}
			if ((sscanf(bufstart, "%s %d %d ", (char *) &bufchar, &packet_number, &buffer_size) == 3)
				&& (strncmp(bufchar, "SJITTER-START", sizeof(bufchar)) == 0) 
				&& (packet_number > 0) && (packet_number <= MAX_PACKET_NUMBER)
				&& (buffer_size >= MIN_BUFFER_SIZE) && (buffer_size <= MAX_BUFFER_SIZE)) {
					checkconf = 1;					
				}
		}
				
		/* Buffer allocation */
		buf = (char *) malloc(buffer_size);
		
		/* 	A START datagram has been received */
		if ((pid = fork()) < 0) {
			errorexit("fork");
		}
		if (pid > 0) {
			/* Father process */
			
			/* Wait the end of the child process */
			waitpid(pid, NULL, 0);
			
			/* Wait for another START datagram */		
			checkconf = 0;
	
		} else { /* pid == 0 */
			/* Child process */
			if (!tag6) {
				/* IPv4 */			
				printf("[Client: %s] Init OK (wait %d datagrams of %d bytes)\n", 
						inet_ntoa(clientaddr.sin_addr), packet_number, buffer_size);
			#ifdef IPV6_SUPPORT
			} else {
				/* IPv6 */
				char buf6[INET6_ADDRSTRLEN];			
				printf("[Client: %s] Init OK (wait %d datagrams of %d bytes)\n", 
						inet_ntop(AF_INET6, &clientaddr6.sin6_addr, buf6, sizeof(buf6)), packet_number, buffer_size);
			#endif
			}
			
			/* Wait data */
			i = 0;
			begin = getcurrenttimems();
			while (!endflag) {
				if (!tag6) {
					/* IPv4 */			
					if (recvfrom(socketfd, buf, buffer_size, 0, (struct sockaddr *) &clientaddr, &slen)==-1)
						errorexit("IPv4 recvfrom error");
				#ifdef IPV6_SUPPORT
				} else {
					/* IPv6 */
					if (recvfrom(socketfd, buf, buffer_size, 0, (struct sockaddr *) &clientaddr6, &slen6)==-1)
						errorexit("IPv6 recvfrom error");					
				#endif
				}
		
				/* Compute Jitter (see in RFC 1889) */
				t_receive = getcurrenttimems();
				if (sscanf (buf, "%s %d %lf ", (char *) &bufchar, &pn, &t_send) > 0) {
					if (strncmp(bufchar, "SJITTER-DATA", sizeof(bufchar)) == 0) {
						/* SJITTER-DATA: the server has received a DATA packet */
						t_transit = t_receive-t_send;
						if (i != 0) {
							t_transitdelta = t_transit - t_transitlast;
							if (t_transitdelta < 0.0)
								t_transitdelta = -t_transitdelta;
							
							/* One way delay */
							onewaydelaylast = t_transit;
							if (onewaydelaylast < onewaydelaymin) 
								onewaydelaymin = onewaydelaylast;
							if (onewaydelaylast > onewaydelaymax) 
								onewaydelaymax = onewaydelaylast;
							onewaydelaysum += onewaydelaylast;	
							onewaydelayavg = onewaydelaysum/i;
							onewaydelaylist[i] = onewaydelaylast;							
														
							/* Jitter */
							jitterlast += (t_transitdelta - jitterlast) / (16.0);
							if (jitterlast < jittermin) 
								jittermin = jitterlast;
							if (jitterlast > jittermax) 
								jittermax = jitterlast;
							jittersum += jitterlast;	
							jitteravg = jittersum/i;
							jitterlist[i] = jitterlast;
							
							if (iflag != 0) {
								/* Only display packets if iglag is set (-i option) */
								if (!tag6) {
									/* IPv4 */			
									printf("[Client: %s] ", inet_ntoa(clientaddr.sin_addr));
								#ifdef IPV6_SUPPORT
								} else {
									/* IPv6 */
									char buf6[INET6_ADDRSTRLEN];			
									printf("[Client: %s] ", inet_ntop(AF_INET6, &clientaddr6.sin6_addr, buf6, sizeof(buf6)));						
								#endif
								}
								printf("One-way delay (ms): Last=%.2lf / Min=%.2lf / Max=%.2lf / Avg=%.2lf\n", onewaydelaylast*1000, onewaydelaymin*1000, onewaydelaymax*1000, onewaydelayavg*1000);
								/*printf("[Client: %s] ", inet_ntoa(clientaddr.sin_addr));*/
								printf("Jitter (ms)       : Last=%.2lf / Min=%.2lf / Max=%.2lf / Avg=%.2lf\n", jitterlast*1000, jittermin*1000, jittermax*1000, jitteravg*1000);
							}
						} else {			
							onewaydelaymin = onewaydelaymax = t_transit;
							jittermin = 65553;						
							if (iflag != 0) {
								if (!tag6) {
									/* IPv4 */			
									printf("[Client: %s] First datagram received\n", inet_ntoa(clientaddr.sin_addr));
								#ifdef IPV6_SUPPORT
								} else {
									/* IPv6 */
									char buf6[INET6_ADDRSTRLEN];			
									printf("[Client: %s] First datagram received\n", inet_ntop(AF_INET6, &clientaddr6.sin6_addr, buf6, sizeof(buf6)));						
								#endif
								}								
							}
						}
						i++;
					} else if (strncmp(bufchar, "SJITTER-END", sizeof(bufchar)) == 0) {
						/* SJITTER-END : the server has received a END packet */
						endflag = 1;
					} else {
						errorexit("Receive an invalid datagram (invalid command)");
					}
				} else {
					errorexit("Receive an invalid datagram (missing parameters))");
				}
				t_transitlast = t_transit;
			}
			end = getcurrenttimems();
							
			/* Display summary */
			if (iflag != 0) {
				if (!tag6) {
					/* IPv4 */			
					printf("[Client: %s] --- Summary\n", inet_ntoa(clientaddr.sin_addr));
				#ifdef IPV6_SUPPORT
				} else {
					/* IPv6 */
					char buf6[INET6_ADDRSTRLEN];			
					printf("[Client: %s] --- Summary\n", inet_ntop(AF_INET6, &clientaddr6.sin6_addr, buf6, sizeof(buf6)));						
				#endif
				}
			}
			if (!tag6) {
				/* IPv4 */			
				printf("[Client: %s] ", inet_ntoa(clientaddr.sin_addr));
			#ifdef IPV6_SUPPORT
			} else {
				/* IPv6 */
				char buf6[INET6_ADDRSTRLEN];			
				printf("[Client: %s] ", inet_ntop(AF_INET6, &clientaddr6.sin6_addr, buf6, sizeof(buf6)));						
			#endif
			}
			printf("Receive %d/%d (%d%%) datagrams in %.2lf seconds (%.0lf Kbps)\n", 
				   i, packet_number, (i*100)/packet_number, end-begin, (packet_number*buffer_size*8)/(end-begin)/1000);
			/*printf("[Client: %s] ", inet_ntoa(clientaddr.sin_addr));*/		   
			printf("One-way delay summary (ms): Min=%.2lf / Max=%.2lf / Avg=%.2lf\n", onewaydelaymin*1000, onewaydelaymax*1000, onewaydelayavg*1000);
			/*printf("[Client: %s] ", inet_ntoa(clientaddr.sin_addr));*/		   
			printf("Jitter summary (ms)       : Min=%.2lf / Max=%.2lf / Avg=%.2lf\n", jittermin*1000, jittermax*1000, jitteravg*1000);
			if (iflag != 0) {
				if (!tag6) {
					/* IPv4 */			
					printf("[Client: %s] --- End\n", inet_ntoa(clientaddr.sin_addr));
				#ifdef IPV6_SUPPORT
				} else {
					/* IPv6 */
					char buf6[INET6_ADDRSTRLEN];			
					printf("[Client: %s] --- End\n", inet_ntop(AF_INET6, &clientaddr6.sin6_addr, buf6, sizeof(buf6)));						
				#endif
				}
			}
													
			exit(0); /* end of the child */
			
			free(buf);
		}

	} /* end of the main loop */
	
	/* Close the socket */
	close(socketfd);
	
	return 0; /* end of the father */

}
